File: ManiacLab/MainMenu.py
# encoding=utf-8
# File name: MainMenu.py
# This file is part of: ManiacLab
#
# LICENSE
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or (at
# your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# FEEDBACK & QUESTIONS
#
# For feedback and questions about ManiacLab please e-mail one of the
# authors named in the AUTHORS file.
########################################################################
from __future__ import print_function, unicode_literals, division
from our_future import *

import Mode
from Engine.UI import ParentWidget, Button, AbstractVBox, LabelWidget, \
    Space, HBox

import Engine.UI.CSS.Minilanguage

class MainMenu(AbstractVBox):
    def __init__(self, parent, **kwargs):
        super(MainMenu, self).__init__(parent, **kwargs)

        title = LabelWidget(self, text="ManiacLab")
        title.StyleClasses.add("title")
        subtitle = LabelWidget(self, text="◀ main menu ▶")
        subtitle.StyleClasses.add("subtitle")


        Space(self),
        Button(self, caption="Select profile"),
        Space(self),
        Button(self, caption="Continue playing"),
        Space(self),
        editor_hbox = HBox(self)
        Button(editor_hbox, caption="Map editor", on_click=self.mapEditor),
        Space(editor_hbox)
        Button(editor_hbox, caption="Tileset editor", on_click=self.tilesetEditor),
        Space(self),
        Button(self, caption="Quit", on_click=self.quit),
        Space(self),

        self.AbsoluteRect.Width = 400
        self.AbsoluteRect.Height = 6 * 30

    def mapEditor(self, sender):
        self.RootWidget.switchMode(self.RootWidget.mapEditorMode)

    def tilesetEditor(self, sender):
        self.RootWidget.switchMode(self.RootWidget.tilesetEditorMode)

    def quit(self, sender):
        self.RootWidget.handleWMQuit()


class Mode(Mode.Mode):
    def __init__(self):
        super(Mode, self).__init__()
        self.desktopWidgets.append(
            MainMenu(None)
            )

Engine.UI.CSS.Minilanguage.ElementNames().register_widget_class(MainMenu)

File: src/game/MainMenu.cpp
/**********************************************************************
File name: MainMenu.cpp
This file is part of: ManiacLab

LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>.

FEEDBACK & QUESTIONS

For feedback and questions about ManiacLab please e-mail one of the
authors named in the AUTHORS file.
**********************************************************************/
#include "MainMenu.hpp"

#include <CEngine/UI/Widgets/RootWidget.hpp>
#include <CEngine/UI/Widgets/LabelWidget.hpp>
#include <CEngine/UI/Widgets/ButtonWidget.hpp>
#include <CEngine/UI/Widgets/SpaceWidget.hpp>

#include "Playground.hpp"
#include "Application.hpp"

using namespace PyEngine;
using namespace PyEngine::UI;

/* MainMenu */

MainMenu::MainMenu():
    AbstractVBox()
{
    LabelWidget *title = new LabelWidget("ManiacLab");
    title->css_classes().insert("title");
    add(title);
    LabelWidget *subtitle = new LabelWidget("◀ main menu ▶");
    subtitle->css_classes().insert("subtitle");
    add(subtitle);
    add(new Space());
    {
        Button *btn_playground = new Button("Playground");
        btn_playground->on_click().connect(
            sigc::mem_fun(this, &MainMenu::playground));
        add(btn_playground);
    }
    add(new Space());
    {
        Button *btn_quit = new Button("Quit");
        btn_quit->on_click().connect(sigc::mem_fun(this, &MainMenu::quit));
        add(btn_quit);
    }
    add(new Space());

    absolute_rect().set_width(400);
    absolute_rect().set_height(6*30);
}

void MainMenu::playground(WidgetPtr sender)
{
    static_cast<Application*>(_root)->switch_to_playground_mode();
}

void MainMenu::quit(WidgetPtr sender)
{
    _root->dispatch_wm_quit();
}

bool MainMenu::is_element(const std::string &name) const
{
    return (name == "mainmenu");
}

/* MainMenuMode */

MainMenuMode::MainMenuMode():
    Mode()
{
    _desktop_widgets.push_back(new MainMenu());
}

void MainMenuMode::enable(Application *root)
{
    Mode::enable(root);
    glClearColor(0, 0.05, 0.1, 1);
}

bool MainMenuMode::ev_key_up(Key::Key key,
                             KeyModifiers modifiers)
{
    switch (key) {
    case Key::q:
    case Key::Escape:
    {
        _root->dispatch_wm_quit();
        return true;
    }
    default: {}
    }
    return false;
}

bool MainMenuMode::ev_wm_quit()
{
    _root->terminate();
    return true;
}

void MainMenuMode::frame_unsynced(TimeFloat deltaT)
{
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    glLoadIdentity();
}
